Tidy writeDynaList()

milux opened the pull request:
### Summary of Changes

Originally, I tried to get rid of writeDynaList() entirely (because of the awful document.writeln()) and do it all in PHP code. I was quite successful, but then I realized that the function was now used from JS code, so instead I cleaned up the code for the one and only place where this function is called now,
The first of the two commits includes excessive explanations how the function was simplified, second one is then just plain code and JSDoc.
### Testing Instructions

none
### Documentation Changes Required

none

milux committed:
Cleaned writeDynaList() in core.js

milux committed:
Removed explanation comments

dgrammatiko commented on media/system/js/core-uncompressed.js:
```diff
@@ -348,42 +348,26 @@ Joomla.editors.instances = Joomla.editors.instances || {};
-		} else {
-			document.writeln( html );
-		}
+		element.innerHTML = html;
```
what will happen if someone call this function without setting the element?

milux replied:
The same as with every other function that is not called according to its definition: It will fail to do its job. In this case, the result is an unchanged DOM.
I removed the else branch because the current implementation is only used in a place where the element is specified, and document.write(ln) is one of the most awful things ever seen in JS, comparable to eval(). Such functions do only cause headache and should imho die out asap.

dgrammatiko replied:
@milux I will totally agree with you but, (in Joomla there is almost always a but) the core.js API cannot change because there might be some 3rd PD that uses it and depends on the `writeln()`, so if that is changed a B/C break is introduced. Same goes for the params. I could have done this when I did https://github.com/joomla/joomla-cms/pull/11040 but unfortunately we can't break B/C in minor versions.
You can rebase this PR against 4.0-dev branch, as the backwards compatibility is not an issue there

milux replied:
Ah, I see... you're right, I didn't consider that this might be used by any 3rd party stuff. (I doubt it, to be honest, but when you guarantee API stability...).
Fine, I will redo anything that changed the public API, some changes might still be useful.

dgrammatiko commented on media/system/js/moduleorder.js:
```diff
@@ -37,7 +37,7 @@
 									$orders[i] = response.data[i].split(',');
 								}
-								writeDynaList('name="' + $name + '" id="' + $id +'"' + $attr, $orders, $originalPos, $originalPos, $originalOrder, $element);
+								writeDynaList('name="' + $name + '" id="' + $id +'"' + $attr, $orders, $originalPos, $originalOrder, $element);
```
why are you dropping one viable here? The function needs 6 vars and you end up using 5? This looks wrong...

milux replied:
Edit: Fixed, see below.

milux committed:
removed all API changes

milux commented:
@dgt41: More happy with the most recent commit?

comp commented:
Production version of core.js broke my site because of writedynalist() and it's use of writeln. Updating to the 4.0 version of core.js fixed this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12184">issues.joomla.org/tracker/joomla-cms/12184</a>.</sub>

milux commented:
Well, most likely something else is broken now, but maybe you won't recognize it...
Are you using some aggressive resource optimizer? Like JCHOptimize with advanced settings?

milux committed:
Merge remote-tracking branch 'upstream/staging' into staging

rdeutz committed:
Merge branch 'staging' into staging

eXsiLe95 commented:
Since this fix is for Joomla 3.x and this version is comming to an end, can this be closed? 
This fix is already open for Joomla 4 on #12282.

Also, this fix did not break my bugtesting installation, but there are no testing instructions provided so I can test.

@iCampus

brianteeman commented:
@eXsiLe95 Joomla 3 has at least three years of supported life remaining

eXsiLe95 commented:
@brianteeman I see the point in Joomla 3.x living for at least the next three years, issues for this version shouldn't be closed right away. Since this issue is "only improving" code (which definitely is important, too), an there is no testing instruction provided, this issue will most probably never get tested (as we can see on the last comment about one year ago.

@milux Can you provide a testing scenario, so that these changes can be tested?

milux commented:
Any use of this function should do the job.
In its latest form, this is nothing more than a small code cleanup.
It should be semantically equivalent. If you can't confirm semantic equivalence by simply comparing the code, I'm comfortable with this being abandoned.

eXsiLe95 commented:
@milux As far as I can see, your code seems to do the same thing, and it clearly reduced the mess in the code. However, only code **looking** like it does the same, it doesn't automatically mean it really **does** the same. Therefore, testing is needed.

Can you tell me where this code is used in Joomla, so that I can test that the changes do not affect the functionalities?

SamuelSchepp commented:
I have tested this item :white_check_mark: successfully on a4759c59ed47ce250ed4bd518192ec242091fa3c<br /><br />Still works as expected.
@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12184">issues.joomla.org/tracker/joomla-cms/12184</a>.</sub>

CiverBlack commented:
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12184">issues.joomla.org/tracker/joomla-cms/12184</a>.</sub>

CiverBlack commented:
I have tested this item :white_check_mark: successfully on a4759c59ed47ce250ed4bd518192ec242091fa3c<br /><br />Works exactly like before

I went in the backend to "Extensions" -> "Modules" and there to one random entry. On the right side you find Ordering. Below that is a dropdown menu. This code runs that. An before and after the patch its still functioning. 

@iCampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12184">issues.joomla.org/tracker/joomla-cms/12184</a>.</sub>

ghost commented:
RTC after two successful tests.

mbabker commented:
The compressed file needs updating too.

ghost commented:
@milux can you please update as @mbabker comment above?

milux committed:
Merge remote-tracking branch 'upstream/staging' into staging

milux committed:
updated compressed core.js

milux commented:
I updated the branch and added a compressed version of the most recent core-uncompressed.js
Is that what you asked for? 🤔 
Why are the compressed versions not automatically generated from the uncompressed ones? That would be far less error-prone imho...

brianteeman commented:
@wilsonge @mbabker any reason that this has not been merged, No pointing in it sitting here for ever

mbabker commented:
I missed the comment about it being updated and apparently scanned over it when looking over RTC lists.  Don't let me forget about it after next week's release.
